source/fsicalmanagement/dataaccess/event_repository.d

```diff
@@ -0,0 +1,259 @@
+{
+    import vibe.data.bson : deserializeBson;
+    import vibe.db.mongo.collection : MongoCollection;
+
+private:
+    @Value("events")
+    MongoCollection events;
+
+public:
+
+    /**
+     * Saves an event to the configured MongoDB collection.
+     * Params:
+     * event = The `Event` to save.
+     * 
+     * Returns: The saved `Event`.
+     */
+    Event save(Event event) @safe
+    {
+        import std.conv : ConvException;
+        import vibe.data.bson : BsonObjectID, serializeToBson;
+
+        try
+        {
+            if (!BsonObjectID.fromString(event.id).valid)
+                throw new ConvException("invalid BsonObjectID.");
+        }
+        catch (ConvException)
+        {
+            event.id = BsonObjectID.generate.to!string;
+        }
+
+        events.insert(event.serializeToBson);
+
+        return event;
+    }
+
+    /**
+     * Gets all events from the configured MongoDB collection.
+     *
+     * Returns: An `InputRange` containing all `Event`s from the configured
+     *          MongoDB collection.
+     */
+    InputRange!Event findAll() @safe
+    {
+        return events.find().map!(deserializeBson!Event).inputRangeObject;
+    }
+
+    /**
+     * Gets an event by its id from the configured MongoDB collection.
+     * Params:
+     * id = The id of the event to get.
+     *
+     * Returns: `Nullable!Event` containing the `Event` corresponding to
+     *          $(D_PARAM id) in the configured MongoDB collection or `null`, if
+     *          no such event exists.
+     */
+    Nullable!Event findById(const string id) @safe
+    {
+        import vibe.data.bson : Bson;
+
+        immutable result = events.findOne(["_id" : id]);
+
+        if (result != Bson(null))
+        {
+            auto event = result.deserializeBson!Event;
+            return event.nullable;
+        }
+        return Nullable!Event.init;
+    }
+
+    /**
+     * Removes an event from the configured MongoDB collection.
+     * Params:
+     * id = The id of the event to remove.
+     */
+    void deleteById(const string id) @safe
+    {
+        events.remove(["_id" : id]);
+    }
+}
+
+/**
+ * A MySQL based implementation of `EventRepository`.
+ */
+class MySQLEventRepository : EventRepository
+{
+private:
+    import mysql.commands : exec, query;
+    import mysql.connection : prepare;
+    import mysql.pool : MySQLPool;
+    import mysql.result : Row;
+
+    MySQLPool pool;
+    @Value("mysql.table.events") string eventsTableName;
+
+public:
+    ///
+    this(MySQLPool pool)
+    {
+        this.pool = pool;
+    }
+
+    /**
+     * Saves an event to the configured MySQL table.
+     * Params:
+     * event = The `Event` to save.
+     * 
+     * Returns: The saved `Event`.
+     */
+    Event save(Event event) @trusted
+    {
+        auto cn = pool.lockConnection();
+        scope (exit)
+            cn.close;
+        auto prepared = cn.prepare("INSERT INTO " ~ eventsTableName
```

Comment:
This seems weird. Create a prepared query on each connection/call?
I'd suppose there is some way to create "named prepared queries" on application launch, and just use those. This should also speed up the database queries.

Reply:
The problem is that these queries depend on the injected `eventsTableName`. Because it is injected with `@Value`, there is no constructor call, in which we could create the prepared statement. I also don't think poodinis supports registering callbacks with `@Value` at the moment. Do you have any idea how to solve this problem?

Reply:
A prepared statement can also receive arguments at call time, like so:

```D
	// Full-featured prepared statements
	Prepared prepared = conn.prepare("SELECT * FROM `tablename` WHERE `name`=? OR `name`=?");
	prepared.setArgs("Bob", "Bobby");
	bobs = conn.query(prepared);
	bobs.close(); // Skip them
```
(taken from the mysql-native readme file)

But I'm not sure if the tablename can be such an argument, and if the actual query has to happen on the same connection the query was prepared with.

Reply:
The query can be on another connection, this is no problem. However, only parameters can be variables in prepared statements, as far as I know.

Reply:
The MySQL-docs don't specify this explicitely. However, they DO have an example on how to select the table dynamically by assigning it to as user variable (see [this docs page](https://dev.mysql.com/doc/refman/8.0/en/sql-syntax-prepared-statements.html), so I suppose this SHOULD be working.

If not, maybe it would be suitable to create some kind of "lazy initialized" prepared statement on class field level?

Reply:
I've addressed this in 83fef61. I'm using lazy initialization for the prepared statements. Additionally, the connections in the connection pool are now being reused. Until now, we closed each connection after executing a prepared statement manually, which is unnecessary.